### PR TITLE
sql: restrict variadic left join lowering #26707

### LINKED ISSUE #26707: Variadic left join lowering results in incorrect local filters
### What version of Materialize are you using?

v0.97.0-dev (70db239b9)

### What is the issue?

The issue occurs when the `enable_variadic_left_join_lowering` system parameter is set to `true`.

Here is a reproduction script:

```sql
CREATE TABLE ft(k INT, v INT NOT NULL);

CREATE TABLE dt(k INT, v INT NOT NULL);

INSERT INTO ft VALUES
  -- one NULL row in ft
  (NULL, 0),
  (1, 1),
  (2, 2),
  (3, 3),
  (4, 4),
  -- mixed row for 5
  (5, 5)
  -- 7 is not present in either table
  ;

INSERT INTO dt VALUES
  (NULL, 0),
  -- 1 not present in dt
  (2, 2),
  (3, 3),
  (4, 4),
  (NULL, 5)
  -- 6 not present in dt
  -- 7 is not present in either table
  ;

-- The predicate on d1: (ft.k = d1.k AND ft.v = d1.k) implies a local predicate
-- on fk: (ft.k = ft.v). However, we should not filter out fk rows where this
-- predicate does not hold.
SELECT
  ft.k AS ft_k,
  ft.v AS ft_v,
  d1.k AS d1_k,
  d1.v AS d1_v,
  d2.k AS d2_k,
  d2.v AS d2_v
FROM
  ft
  LEFT JOIN dt AS d1 ON (ft.k = d1.k AND ft.v = d1.k)
  LEFT JOIN dt AS d2 ON (ft.k = d2.v);
```

This produces the following 5 rows in Materialized

```text
 ft_k | ft_v | d1_k | d1_v | d2_k | d2_v 
------+------+------+------+------+------
    1 |    1 |      |      |      |     
    5 |    5 |      |      |      |    5
    2 |    2 |    2 |    2 |    2 |    2
    3 |    3 |    3 |    3 |    3 |    3
    4 |    4 |    4 |    4 |    4 |    4
(5 rows)
```

in Materialize, but 6 rows in Postgres

```text
 ft_k | ft_v | d1_k | d1_v | d2_k | d2_v 
------+------+------+------+------+------
    1 |    1 |      |      |      |     
    2 |    2 |    2 |    2 |    2 |    2
    3 |    3 |    3 |    3 |    3 |    3
    4 |    4 |    4 |    4 |    4 |    4
    5 |    5 |      |      |      |    5
      |    0 |      |      |      |     
(6 rows)
```

The

```text
      |    0 |      |      |      |     
```

is missing. Looking at the `EXPLAIN LOCALLY OPTIMIZED PLAN` for this statement we can see that because `ft.k = d1.k AND ft.v = d1.k` implies `ft.k = ft.v`, we have a local `Filter (#0{k} = #1{v})` on top of `Get materialize.public.ft` which should not be there.

```
Return
  Project (#0, #1, #9..=#12)
    Map ((#4) IS NULL, case when #8 then null else #0{k} end, case when #8 then null else #3{v} end, case when (#7) IS NULL then null else #5{k} end, case when (#7) IS NULL then null else #0{k} end)
      Join on=(#0{k} = #2{k} = #6{v})
        Filter (#0{k} = #1{v})
          Get materialize.public.ft
        Union
          Map (true)
            Get l0
          Project (#1..=#3)
            Map (null, null)
              Threshold
                Union
                  Project (#0, #0)
                    Negate
                      Get l0
                  Distinct project=[#1{v}, #0{k}]
                    Get materialize.public.ft
        Union
          Map (true)
            Get materialize.public.dt
          Project (#1, #0, #2)
            Map (null, null)
              Threshold
                Union
                  Project (#1)
                    Negate
                      Get materialize.public.dt
                  Distinct project=[#0{k}]
                    Get materialize.public.ft
With
  cte l0 =
    Filter (#0{k}) IS NOT NULL
      Get materialize.public.dt

Target cluster: quickstart
```

### PR DESCRIPTION
**Pull request by aalexandrov**
This is a fix for #26707 that works by further restricting the set of supported `ON` predicates. In particular, we skip variadic left join lowering for predicates of the form `lhs.x  = rhs.z AND lhs.y = rhs.z` which imply `x = y`.

### Motivation

  * This PR fixes a recognized bug.

Fixes #26707.

### Tips for reviewer

The fix is in the following commit

* sql: prevent `variadic_left` from pushing incorrect lhs filters

The first commit adds some more comments to `variadic_left.rs`, and the remaining commits add `*.slt` and `*.rqg` tests (those have been reviewed by @def- already).

To validate a proposed fix, run

```bash
RUST_BACKTRACE=full bin/sqllogictest -- test/sqllogictest/github-26707.slt
```

and

```bash
bin/mzcompose --find rqg run default left-join-stacks --queries 1000 --duration 3600
```

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by aalexandrov** [comment]
@MaterializeInc/testing : I'm not sure how to (and in general whether we should) add the new RQG workload to the Nightly runs.

File: test/rqg/mzcompose.py [hunk 1]
```diff
@@ -127,19 +134,36 @@ class Workload:
         reference_implementation=None,
         validator="QueryProperties,RepeatableRead",
     ),
+    # Added as part of MaterializeInc/materialize#25340.
+    Workload(
+        name="left-join-stacks",
```
Comment (def-) [inline test/rqg/mzcompose.py:139]: Please add a new rqg test in ci/nightly/pipeline.template.yml using `left-join-stacks`, then we will run it in Nightly automatically too.

**Review by def- — CHANGES_REQUESTED** [review]
I'm probably less familiar with RQG than you are. I'm happy with having both a deterministic reproducer and a randomized one (RQG scenario)

The code changes seem good to me other than the one thing I marked inline.

File: test/rqg/mzcompose.py (hunk 1 above)
Comment (aalexandrov) [reply test/rqg/mzcompose.py:139]: Will do. I'll probably use shorter `duration` and `queries` for this workload, though (I think the current grammar doesn't open a very big query space, so running `1000` queries should be enough to catch regressions (the current one is caught with 10 queries for example).

File: test/rqg/mzcompose.py [hunk 2]
```diff
@@ -149,7 +178,12 @@ def workflow_default(c: Composition, parser: WorkflowArgumentParser) -> None:
     parser.add_argument(
         "--duration",
         type=int,
-        help="Run the Workload for the specifid time in seconds",
+        help="Run the Workload for the specified time in seconds",
+    )
+    parser.add_argument(
```
Comment (aalexandrov) [inline test/rqg/mzcompose.py:183]: Added `--queries` CLI option for directly providing a the number of queries while developing / debugging.

File: test/rqg/mzcompose.py [hunk 3]
```diff
@@ -193,40 +227,35 @@ def workflow_default(c: Composition, parser: WorkflowArgumentParser) -> None:
...
-            else None,
+            image=materialize_image(args.this_tag),
             use_default_volumes=False,
         ),
     ]
 
+    # A list of psql URLs for dataset initialization
     psql_urls = ["postgresql://materialize@mz_this:6875/materialize"]
 
-    if args.other_tag == "common-ancestor":
```
Comment (aalexandrov) [inline test/rqg/mzcompose.py:209]: Moved this to a `StoreOtherTag(argparse.Action)` implementation (used with `action=StoreOtherTag` when configuring the `argparse` argument).

**Review by def- — APPROVED** [review]

File: src/sql/src/plan/lowering/variadic_left.rs [hunk 4]
```diff
@@ -389,6 +399,19 @@ fn decompose_equations(predicate: &MirScalarExpr) -> Option<Vec<(usize, usize)>>
             _ => return None,
         }
     }
+
+    // Remove duplicates
```
Comment (aalexandrov) [inline src/sql/src/plan/lowering/variadic_left.rs:403]: This extension of `decompose_equations` should fix the issue (the regression test doesn't fail). Once I have a release build I will also try RQG.

**Review by frankmcsherry — APPROVED** [review]
The fix looks good, and most of the rest of the stuff I do not grok!

**Pull request merged**